### PR TITLE
Two new documentation notebooks

### PR DESCRIPTION
**Pull request by aaron-siegel**
Two new documentation notebooks:
- Queries and Expressions (section 3 in the Fundamentals tutorial)
- working-with-llama-cpp

### CODE REVIEW

**Comment by aaron-siegel** [comment]
This change is [Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/349)

File: pixeltable/catalog/table.py
Comment (pierrebrunelle) [inline pixeltable/catalog/table.py:1]: I'd merge the Notes and Best Practices at the bottom with the Important Notes section at the top as it's now redundant. Apart from that all good.

File: docs/release/fundamentals/queries-and-expressions.ipynb
Comment (pierrebrunelle) [inline docs/release/fundamentals/queries-and-expressions.ipynb:1]: **I meant to comment on the Llama.cpp notebook here..**

I'd merge the Notes and Best Practices at the bottom with the Important Notes section at the top as it's now redundant. Apart from that all good.

Comment (pierrebrunelle) [reply docs/release/fundamentals/queries-and-expressions.ipynb:1]: I meant to comment on the Llama.cpp notebook here..

Comment (aaron-siegel) [reply docs/release/fundamentals/queries-and-expressions.ipynb:1]: Done.

**Review by pierrebrunelle — APPROVED** [review]
Approving my comment, but will let others review the queries and expressions tutorial.

File: docs/release/fundamentals/queries-and-expressions.ipynb [hunk 1]
```diff
@@ -0,0 +1,2023 @@
...
+   "source": [
+    "### Column References"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1cd19032-d232-4c38-82eb-b4246cf712ca",
+   "metadata": {},
+   "source": [
+    "The most basic type of expression is a __column reference__: that's what you get when you type, say, `t.image`. An expression such as `t.image` by itself is just a Python object; it doesn't contain any actual data, and no data will be loaded until you use the expression in a `select()` query or `add_column()` statement. Here's what we get if we type `t.image` by itself:"
```
Comment (mkornacker) [inline docs/release/fundamentals/queries-and-expressions.ipynb:212]: To be discussed: should describe() on a columnref output a table? Maybe something more streamlined, similar to how a series in pandas is displayed differently than a dataframe.

File: docs/release/fundamentals/queries-and-expressions.ipynb [hunk 2]
```diff
@@ -0,0 +1,2023 @@
...
+   "source": [
+    "t"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9bfc6d8e-9a61-4399-8689-974caef19b24",
+   "metadata": {},
+   "source": [
+    "Pixeltable provides a range of operators on `Json`-typed output that behave just as you'd expect. To look up a key in a dictionary, use the syntax `t.classification['labels']`:"
```
Comment (mkornacker) [inline docs/release/fundamentals/queries-and-expressions.ipynb:456]: Worth mentioning that these are json path expressions, ie, we didn't just make this up randomly, and providing a link to something like
https://en.wikipedia.org/wiki/JSONPath

File: docs/release/fundamentals/queries-and-expressions.ipynb [hunk 3]
```diff
@@ -0,0 +1,2023 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "73d8eb1c-d449-4990-b643-4e4e729fd29f",
+   "metadata": {},
+   "source": [
+    "The \"attribute\" syntax isn't fully general (it won't work for dictionary keys that are not valid Python identifiers), but it's handy when it works.\n",
+    "\n",
+    "\n",
+    "`t.classification.labels` is another Pixeltable expression; you can think of it as saying, \"do the `'labels'` lookup from every dictionary in the column `t.classification`, and return the result as a new column.\" As before, the expression by itself contains no data; it's the query that does the actual work of retrieving data. Here's what we see if we just give the expression by itself, without a query:"
```
Comment (mkornacker) [inline docs/release/fundamentals/queries-and-expressions.ipynb:563]: Maybe we should implement Expr.__repr__() (now) to print ColumnType.__str__().

I was expecting this to invoke Expr.__str__().

File: docs/release/fundamentals/queries-and-expressions.ipynb [hunk 4]
```diff
@@ -0,0 +1,2023 @@
...
+    "\n",
+    "t.select(t.image, custom_dict).head(5)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8377bb7f-f665-4ff9-a694-58c93ae28583",
+   "metadata": {},
+   "source": [
+    "### Function Calls\n",
```
Comment (mkornacker) [inline docs/release/fundamentals/queries-and-expressions.ipynb]: I'm wondering if 'Function Calls' is confusing and if UDF Calls would be better.

File: docs/release/fundamentals/queries-and-expressions.ipynb [hunk 5]
```diff
@@ -0,0 +1,2023 @@
...
+   "source": [
+    "t.select(t.image, t.image.localpath).head(5)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e76aaa1f-5a13-46f4-a507-26e6a7729512",
+   "metadata": {},
+   "source": [
+    "Any computed column will have two additional properties, `errortype` and `errormsg`. These properties will usually be `None`. However, if the computed column was created with `on_error='ignore'` and an exception was encountered during column execution, then the properties will contain additional information about the exception.\n",
```
Comment (mkornacker) [inline docs/release/fundamentals/queries-and-expressions.ipynb:1659]: Also link to the Working with External Files tutorial.

File: docs/release/fundamentals/queries-and-expressions.ipynb [hunk 6]
```diff
@@ -0,0 +1,2023 @@
...
+   "source": [
+    "t.select(t.image, t.channel, t.channel.errortype, t.channel.errormsg).head(5)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6847abcb-1b78-4031-b097-8f4b7263b192",
+   "metadata": {},
+   "source": [
+    "## The Pixeltable Type System"
```
Comment (mkornacker) [inline docs/release/fundamentals/queries-and-expressions.ipynb:1854]: I think this needs to be in the first of the fundamentals tutorials. That would be the very first question on my mind if I learn about a new database system: how do I define tables and what are the supported types.

File: docs/release/fundamentals/queries-and-expressions.ipynb [hunk 7]
```diff
@@ -0,0 +1,2023 @@
...
+    "| pxt.Array | np.ndarray |\n",
+    "| pxt.Image | PIL.Image.Image |\n",
+    "| pxt.Video | str |\n",
+    "| pxt.Audio | str |\n",
+    "| pxt.Document | str |\n",
+    "\n",
+    "The Python type is what you'll get back if you query an expression of the given Pixeltable type. For `pxt.Json`, it can be any of `str`, `int`, `float`, `bool`, `list`, or `dict`.\n",
+    "\n",
+    "<div class=\"alert alert-block alert-info\">\n",
+    "<code>pxt.Audio</code>, <code>pxt.Video</code>, and <code>pxt.Document</code> all correspond to the Python type <code>str</code>. This is because those types are represented by URLs or file paths that reference the media in question.\n",
```
Comment (mkornacker) [inline docs/release/fundamentals/queries-and-expressions.ipynb]: You should point out that if you do t.select(t.audio_column) you'll get back the local path, not the url you put in there. The assumption is that you'll want to do something with the actual data.

File: docs/release/fundamentals/queries-and-expressions.ipynb [hunk 8]
```diff
@@ -0,0 +1,2023 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ff20bf54-f4a0-435b-9874-13a4e8471dd3",
+   "metadata": {},
+   "source": [
+    "Several types can be __specialized__ to constrain the allowable data in a column.\n",
+    "\n",
+    "- `pxt.Image` can be specialized with a resolution and/or an image mode.\n",
```
Comment (mkornacker) [inline docs/release/fundamentals/queries-and-expressions.ipynb]: I think this should be concrete (ie, how exactly do I do that?)

File: docs/release/tutorials/working-with-llama-cpp.ipynb [hunk 9]
```diff
@@ -0,0 +1,1603 @@
...
+    "- Consider memory usage when choosing models and quantizations"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "mPfFuWBgZJ5q"
+   },
+   "source": [
+    "## Setup Environment\n",
```
Comment (mkornacker) [inline docs/release/tutorials/working-with-llama-cpp.ipynb]: Pet peeve of mine:
'setup': noun
'set up': verb

File: docs/release/tutorials/working-with-llama-cpp.ipynb [hunk 10]
```diff
@@ -0,0 +1,1603 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "-hdc6e78ZJ5r"
+   },
+   "source": [
+    "## Create a Table for Chat Completions\n",
+    "\n",
+    "Now let's create a table that will contain our inputs and responses."
```
Comment (mkornacker) [inline docs/release/tutorials/working-with-llama-cpp.ipynb:61]: Would be good to link to other docs pages. In this case, the tutorial on creating tables.

In general, I think we should assume that first-time users might start out by reading one of these applied tutorials, and we can use them as opportunities to provide additional links for further reading.

File: docs/release/tutorials/working-with-llama-cpp.ipynb [hunk 11]
```diff
@@ -0,0 +1,1603 @@
...
+    "pxt.create_dir('llama_demo')\n",
+    "\n",
+    "t = pxt.create_table('llama_demo.chat', {'input': pxt.String})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, we add a computed column that calls the Pixeltable `create_chat_completion` UDF, which adapts the corresponding llama.cpp API call. In our examples, we'll use pretrained models from the Hugging Face repository. llama.cpp makes it easy to do this by specifying a repo_id (from the URL of the model) and filename from the model repo; the model will then be downloaded and cached automatically.\n",
```
Comment (mkornacker) [inline docs/release/tutorials/working-with-llama-cpp.ipynb:93]: Same for udfs.

File: docs/release/tutorials/working-with-llama-cpp.ipynb [hunk 12]
```diff
@@ -0,0 +1,1603 @@
...
+       "UpdateStatus(num_rows=3, num_computed_values=9, num_excs=0, updated_cols=[], cols_with_excs=[])"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Test with a simple question\n",
```
Comment (mkornacker) [inline docs/release/tutorials/working-with-llama-cpp.ipynb:199]: This isn't llama-cpp-specific anymore. Should we use this as a template for every llm api? Should we have a combined llm-centric tutorial that shows common operations (and then maybe show code specific to each particular api in tabs)?

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/349)* status: 0 of 16 files reviewed, 13 unresolved discussions (waiting on @mkornacker and @pierrebrunelle)

___
*[`docs/release/fundamentals/queries-and-expressions.ipynb` line 456 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/349#gh-1829887017:-OB2klOq7S7n4NhVxLzW:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/2b265a86dea5ff946653073f031ddc1acce53847/docs/release/fundamentals/queries-and-expressions.ipynb#L456)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Worth mentioning that these are json path expressions, ie, we didn't just make this up randomly, and providing a link to something like
https://en.wikipedia.org/wiki/JSONPath

</blockquote></details>

Done.
___
*[`docs/release/fundamentals/queries-and-expressions.ipynb` line 563 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/349#gh-1829895731:-OB2kx1q0ygYBX4Drfov:be5435t) ([raw file](https://github.com/pixeltable/pixeltable/blob/2b265a86dea5ff946653073f031ddc1acce53847/docs/release/fundamentals/queries-and-expressions.ipynb#L563)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Maybe we should implement Expr.__repr__() (now) to print ColumnType.__str__().

I was expecting this to invoke Expr.__str__().
</blockquote></details>

Agreed. We should think through the `__repr__()` output of all the user-facing classes (i.e., all the classes a user might `__repr__()` at some point). Maybe as part of our "Soft Launch" goal. I'll create a ticket for this
___
*[`docs/release/fundamentals/queries-and-expressions.ipynb` line 929 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/349#gh-1829897933:-OB2lQzV1Bzo8YeZz6pF:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/2b265a86dea5ff946653073f031ddc1acce53847/docs/release/fundamentals/queries-and-expressions.ipynb#L929)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

I'm wondering if 'Function Calls' is confusing and if UDF Calls would be better.
</blockquote></details>

Done.
___
*[`docs/release/fundamentals/queries-and-expressions.ipynb` line 1659 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/349#gh-1829904852:-OB2lccYCDLPJiLaWL5b:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/2b265a86dea5ff946653073f031ddc1acce53847/docs/release/fundamentals/queries-and-expressions.ipynb#L1659)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Also link to the Working with External Files tutorial.
</blockquote></details>

Done.
___
*[`docs/release/fundamentals/queries-and-expressions.ipynb` line 1846 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/349#gh-1829907189:-OB2uU9b-7zJqgOOiTQf:b-blgt2l) ([raw file](https://github.com/pixeltable/pixeltable/blob/2b265a86dea5ff946653073f031ddc1acce53847/docs/release/fundamentals/queries-and-expressions.ipynb#L1846)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

I think this needs to be in the first of the fundamentals tutorials. That would be the very first question on my mind if I learn about a new database system: how do I define tables and what are the supported types.
</blockquote></details>

Ok. What if I give the list of supported types and their Python equivalents in the first tutorial, but defer the detailed discussion of specialized types, error handling, etc. until this one? I think it'll flow better, and the repetition will actually have value for something this important. (I'd include a note & link in the first tutorial saying "We'll discuss types in greater depth in the Queries & Expression section")
___
*[`docs/release/fundamentals/queries-and-expressions.ipynb` line 1875 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/349#gh-1829909137:-OB2swHz252mUjKHigZA:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/2b265a86dea5ff946653073f031ddc1acce53847/docs/release/fundamentals/queries-and-expressions.ipynb#L1875)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

You should point out that if you do t.select(t.audio_column) you'll get back the local path, not the url you put in there. The assumption is that you'll want to do something with the actual data.
</blockquote></details>

Done.
___
*[`docs/release/fundamentals/queries-and-expressions.ipynb` line 1886 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/349#gh-1829910247:-OB2uT1pAoyw-QRQitw1:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/2b265a86dea5ff946653073f031ddc1acce53847/docs/release/fundamentals/queries-and-expressions.ipynb#L1886)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

I think this should be concrete (ie, how exactly do I do that?)
</blockquote></details>

Done.

**Pull request merged**